### PR TITLE
Update launcher docs

### PR DESCRIPTION
**Pull request by SethBarberee**
Went peeking around some of the widget code and decided to update it a bit. Still messing with launcher functionality so I might have other examples coming soon but this one works pretty well for now.

Feedback appreciated!

### CODE REVIEW

**Comment by Elv13** [comment]
Yay! More doc examples, thanks for contributing those, new users love them.

The CI fails because of:

```

Checking tests/examples/awful/widget/launcher/default.lua 1 warning

    tests/examples/awful/widget/launcher/default.lua:21:32: accessing undefined variable 'hotkeys_popup'

Checking tests/examples/awful/widget/layoutlist/bar.lua OK
```

This can be fixed by adding `local hotkeys_popup = nil --DOC_HIDE` in the example. It is never accessed, so it will be fine.

**Comment by SethBarberee** [comment]
Fixed. Thanks for the tip.

File: lib/awful/widget/launcher.lua [hunk 1]
```diff
@@ -12,8 +16,10 @@ local button = require("awful.button")
 local launcher = { mt = {} }
 
 --- Create a button widget which will launch a command.
--- @param args Standard widget table arguments, plus image for the image path
--- and command for the command to run on click, or either menu to create menu.
+-- @tparam table args
+-- @tparam string args.image image path to an image
+-- @tparam string args.command command to run on click
```
Comment (Aire-One) [inline lib/awful/widget/launcher.lua]:
```suggestion
-- @tparam string args.command Command to run on user click.
```

File: lib/awful/widget/launcher.lua [hunk 2]
```diff
@@ -12,8 +16,10 @@ local button = require("awful.button")
...
+-- @tparam table args.menu table of items to create a menu
```
Comment (Aire-One) [inline lib/awful/widget/launcher.lua]:
```suggestion
-- @tparam table args.menu Table of items to create a menu based on `awful.menu`.
```

File: lib/awful/widget/launcher.lua [hunk 3]
```diff
@@ -12,8 +16,10 @@ local button = require("awful.button")
 local launcher = { mt = {} }
 
 --- Create a button widget which will launch a command.
--- @param args Standard widget table arguments, plus image for the image path
--- and command for the command to run on click, or either menu to create menu.
+-- @tparam table args
+-- @tparam string args.image image path to an image
```
Comment (Aire-One) [inline lib/awful/widget/launcher.lua]:
```suggestion
-- @tparam image args.image The image to display on the launcher button (refer to `wibox.widget.imagebox`).
```

File: lib/awful/widget/launcher.lua [hunk 4]
```diff
@@ -12,8 +16,10 @@ local button = require("awful.button")
...
+-- @tparam string args.command command to run on click
+-- @tparam table args.menu table of items to create a menu
 -- @return A launcher widget.
 -- @constructorfct awful.widget.launcher
```
Comment (Aire-One) [inline lib/awful/widget/launcher.lua]: I'd add references to help user understanding arguments:
```suggestion
-- @constructorfct awful.widget.launcher
-- @see wibox.widget.imagebox
-- @see awful.menu
```

File: lib/awful/widget/launcher.lua [hunk 5]
```diff
@@ -12,8 +16,10 @@ local button = require("awful.button")
 local launcher = { mt = {} }
 
 --- Create a button widget which will launch a command.
--- @param args Standard widget table arguments, plus image for the image path
--- and command for the command to run on click, or either menu to create menu.
+-- @tparam table args
+-- @tparam string args.image image path to an image
+-- @tparam string args.command command to run on click
+-- @tparam table args.menu table of items to create a menu
 -- @return A launcher widget.
```
Comment (Aire-One) [inline lib/awful/widget/launcher.lua]: I'd also add the widget type for the return:
```suggestion
-- @treturn awful.widget.launcher A launcher widget.
```

File: lib/awful/widget/launcher.lua (hunk 2 above)
Comment (Aire-One) [inline lib/awful/widget/launcher.lua]: From the function first line:
```lua
if not args.command and not args.menu then return end
```

So both arguments are optional but at least one is mandatory to the function run.

I'd use `-- @tparam[opt]` for both and stipulate in the function description that it require one of them for the function to run.

Comment (SethBarberee) [reply lib/awful/widget/launcher.lua]: Agreed. I'll fix this up later today

**Comment by Elv13** [comment]
Thanks for the update. The CI failed because of:

```
Checking lib/awful/widget/launcher.lua            1 warning

lib/awful/widget/launcher.lua:2:121: line is too long (123 > 120)
```

It is actually important because the first line of the module documentation is used in the "welcome page" index. It should fit in a single line on 720p (cheap laptop still have 720p-ish 1366x768) screens with the default font size.

**Comment by Aire-One** [comment]
> It should fit in a single line on 720p (cheap laptop still have 720p-ish 1366x768) screens with the default font size.

I didn't know this was a requirement... sorry.

What about:

```lua
-- A button widget which hosts a menu or starts a command.
```

The detailed explanation should go to the description.

**Comment by actionless** [comment]
please squash/rebase

**Comment by SethBarberee** [comment]
Applied comments and squashed

File: lib/awful/widget/launcher.lua [hunk 6]
```diff
@@ -1,4 +1,12 @@
 ---------------------------------------------------------------------------
+--  A button widget which hosts a menu or starts a command
```
Comment (Aire-One) [inline lib/awful/widget/launcher.lua]: Don't miss the final dot `.` on summaries. It breaks ldoc.
```suggestion
--  A button widget which hosts a menu or starts a command.
```

File: lib/awful/widget/launcher.lua [hunk 7]
```diff
@@ -1,4 +1,12 @@
...
+--
+--  Implementation of `awful.widget.button` with the ability to host an `awful.menu` or a command to execute on user click.
```
Comment (actionless) [inline lib/awful/widget/launcher.lua]:
```suggestion
--  Implementation of `awful.widget.button` with the ability to host an `awful.menu`
--  or a command to execute on user click.
```

Comment (Elv13) [reply lib/awful/widget/launcher.lua]: Done

File: lib/awful/widget/launcher.lua (hunk 6 above)
Comment (Elv13) [reply lib/awful/widget/launcher.lua]: Done

**Comment by Elv13** [comment]
Thanks for this. Sorry for the friction. LDoc and luacheck are tricky. Either you know everything about them or they do unexpected things. Hopefully the next time it will be smoother given you know those pitfalls now.

I really appreciate these kind of contributions. Documentation is one area where the maintainers are not the right people to do the job. We know the internals and tend to document well things nobody care about while obvious things remain undocumented and we don't see them because we don't *use* the doc.

**Pull request merged**